test/pmtu.py

```diff
@@ -0,0 +1,91 @@
+import argparse
+import ipaddress
+from icmplib import ping
+
+
+def findmtu(host, verbose=False, debug=False):
+    """
+    Find the PMTU to the specified host.
+    Searches the range
+    Host is an IPv4 or IPv6 address.
+    """
+
+    # These headers get added to the ping payload to get to MTU
+    SIZE_ICMP_HDR = 8
+    SIZE_IPV4_HDR = 20
+    SIZE_IPV6_HDR = 40
+
+    # These Ethernet headers are not included in the MTU.
+    # Subtract them from the size-on-wire to get MTU.
+    SIZE_ETH2_HDR = 14
+
+    if isinstance(ipaddress.ip_address(host), ipaddress.IPv6Address):
+        size_ip_hdr = SIZE_IPV6_HDR
+    else:
+        size_ip_hdr = SIZE_IPV4_HDR
+
+    lower = 900
+    upper = 10000
+    MTU = lower
+
+    if verbose:
+        print(f"\nhost: {host}")
+
+    result = ping(host)
+    if not result.is_alive:
+        return 0
+
+    timeout = 1.5 * result.max_rtt / 1E3
+    interval = 0.5 * result.max_rtt / 1E3
+    if verbose:
+        print(result)
+        print(f'PMTU discovery timeout: {timeout}, interval: {interval}')
+
+    test_size = 1700
+    while upper - lower > 1:
+        test_MTU = test_size + SIZE_ICMP_HDR + size_ip_hdr
+        if verbose:
+            if debug:
+                s = f" - MTU search range: {upper}-{lower} - {test_MTU+SIZE_ETH2_HDR} bytes on wire "
+            else:
+                s = ""
+            print(
+                f"checking frame size {test_MTU} {s}",
+                end="",
+                flush=True,
+            )
+        result = ping(
+            host,
+            count=2,
+            interval=interval,
+            timeout=timeout,
+            payload_size=test_size,
+            pmtudisc_opt="do",
+        )
+        if result.is_alive:
+            lower = test_size
+            MTU = test_MTU
+            if verbose:
+                print("✅")
+        else:
+            upper = test_size
+            if verbose:
+                print("❌")
+
+        test_size = lower + int((upper - lower) / 2)
```

Comment:
Thanks for implementing this feature. I'm currently looking into adding PMTUD to MeshPing, a little network monitoring tool I'm building. I'm currently working on a somewhat-hacky PMTUD implementation, but I'd also love to contribute back to icmplib, hence I found your PR. Your implementation looks a lot cleaner than mine :smile: I'd just like to point out that, rather than guessing and trying new MTU values when you get an error back, the new MTU is actually included in the downstream gateways' ICMP messages, and you can [query them from the socket](https://github.com/Svedrin/meshping/blob/c35bf1363b8d722ab42abd4e5f527904dc48d694/src/socklib.py#L38-L39):

```
    def get_mtu(self):
        return self._sock.getsockopt(socket.IPPROTO_IP, IP_MTU)
```

Similar [for IPv6](https://github.com/Svedrin/meshping/blob/c35bf1363b8d722ab42abd4e5f527904dc48d694/src/socklib.py#L55-L56):

```
    def get_mtu(self):
        return self._sock.getsockopt(socket.IPPROTO_IPV6, IPV6_MTU)
```

Maybe this would be a nice addition? :slightly_smiling_face:

Reply:
Thanks for the comment and apologies for the delayed response.

I agree, a brute-force PMTU discovery mechanism like this is not often needed these days.

However, the reason I had for implementing this feature is that there are some network devices, like fibre media converters, that operate as bridges on an Ethernet segment, and are entirely transparent at the IP layer - except when they're dropping large Ethernet frames.

In that scenario, there are no ICMP messages to signal the MTU constraint, and that's the scenario I needed this tool for.